### ISSUE TITLE
Split Shims and Functional Core

### ISSUE THREAD

**Issue by kitsonk** [dojo/core#145]
@novemberborn pointed out that [`dojo/compose`](/dojo/compose) currently only has a dependency of `WeakMap` on core.  It raised the thought in my head that maybe we should split core into things that provide functional shims to ES6+ (and W3C APIs not fully available) and things that are actual functional enhancements to APIs that we would prefer to use (e.g. `request`).  This could reduce dependendicies for some packages and give a clear distinction between replicating APIs and providing _unique_ APIs.  `Task` would go into _core_ while `Promise` would be in _shims_.

It would also reduce my _worry_ about Streams being in core.  That would clearly go into the shims package, because it is trying to replicate an future API.  It also gives us a clear option to produce a "modern" version of the shims package that is essentially nothing but `export default Promise;`.

**Comment by novemberborn** [comment]
👍 

Even unique APIs could be in their own package though.

**Comment by dylans** [comment]
> Even unique APIs could be in their own package though.

I agree with the plan for splitting out the shims, as I think it will make it much more natural to drop each shim when it is no longer needed.

I do think we want to be careful to consider resisting the urge to split every API into their own package (I know that is often the preferred approach for Node.js devs). I think our reasoning was that ES modules do make it fairly easy to import and export just parts of packages, core itself is pretty small, and we wanted to avoid having 10 separate packages that all depend on one other.

I think it's easy enough to decide if there are things that are fairly big in core besides shims that should be split out from core, especially as the dependency chain becomes clearer.

**Comment by tomdye** [comment]
Sounds good

**Comment by novemberborn** [comment]
> I think it's easy enough to decide if there are things that are fairly big in core besides shims that should be split out from core, especially as the dependency chain becomes clearer.

Yes.

> I do think we want to be careful to consider resisting the urge to split every API into their own package (I know that is often the preferred approach for Node.js devs). I think our reasoning was that ES modules do make it fairly easy to import and export just parts of packages, core itself is pretty small, and we wanted to avoid having 10 separate packages that all depend on one other.

Guilty! 😄  That said, separating functionality into different packages enables you to evolve each of these packages towards stability. Once `core` is at `2.0.0`, how would you start iterating on a new addition (including breaking changes)? You'll probably have to do development in a separate package and only once that becomes stable merge it into core.

**Comment by kitsonk** [comment]
> Once core is at 2.0.0, how would you start iterating on a new addition (including breaking changes)?

While we haven't detailed it out, I think it is reasonable to believe that additional APIs could be added prior to `<3.0.0`, just no breaking existing APIs.  I hope/expect that once we break the back of the emotionally loaded Dojo 2, we will start iterating quickly and each package taking its own trajectory.  But someone would be able to expect to depend on the first version of `^2.0.0` that has the feature they are looking for.

Take a perfect example, if Core 2 was released, I would still have no objection to integrating the Fetch API into `core/request`.  On the other hand, if we were going to say, only provide an ES Observable interface to `core/request` and get rid of the `Promise` returns, then we wouldn't do that until Core 3.

**Comment by novemberborn** [comment]
> if Core 2 was released, I would still have no objection to integrating the Fetch API into core/request.

Indeed. But once you've integrated the Fetch API you can't then break it until Core 3. And you'll have to iterate on it in pre-release versions. But you can't merge these in-progress APIs into `master` because then you can't cut normal releases easily, which means each new API gets developed in its own branch and either has its own pre-release (`2.1.3-fetch-api.4`) or each week you do an integration branch of everything that's in progress.

That nightmare can be solved by incubating new APIs in separate packages and then merging them into Core, and then you can do a minor Core release to add features. But if you're doing that why not keep them as separate packages to begin with?

I suppose Core could depend on all those separate packages (using exact version pins) and bump its version based on the underlying changes. Though then you may get stuck with peer dependencies again. I guess that's why lodash arrived at the solution they did.

**Comment by kitsonk** [comment]
> But once you've integrated the Fetch API you can't then break it until Core 3.

Yes, but I don't think that would be 6 years in the future.

> I guess that's why lodash arrived at the solution they did.

Which is horribly inefficient and obtuse and artificially makes it the most used _package_ on npm.  It is like a thousand `left-pad`s.  Most of the APIs in `core` are not new and actually not iterated much upon and are foundational.

If we have a build system (see dojo/meta#10) that meets its functional requirements the dependency management becomes easier.  We should strike a balance between `lodash` and DojoX.  I still think using `npm` and those dependencies as the way to only include the code you need in a consumable application is a fools errand.

After a discussion with @dylans, @kfranqueiro and others, it made me realise that the functional enhancements in `core` and the standards based shims are not easily separable.  We might want to give it a bit more thought and see if the patterns we use in the standards based shims could be further reduced to have no dependency on `core` and `core` could build upon those shims.  I think that _might_ be possible.  The biggest "external" dependency in the shims are `global` and the `hasClass` decorator which we use for applying shims.  There are a couple of other dependencies, but they look like they are other shims.  `Promise` uses `queueMicroTask` from `queue`.  That would in theory be a minor inefficiency.

**Comment by novemberborn** [comment]
> I still think using npm and those dependencies as the way to only include the code you need in a consumable application is a fools errand.

Could you elaborate? I think npm is useful for managing dependencies. The build system would then do tree shaking and whatnot to produce runnable code.

I appreciate that we don't want to ask people to install 20 packages in order to build a Dojo 2 application, but then why even have `dojo/compose` outside of `dojo/core`?

**Comment by kitsonk** [comment]
> Could you elaborate? I think npm is useful for managing dependencies. The build system would then do tree shaking and whatnot to produce runnable code.

It does, if you know what those dependencies are, where you should locate them and more specifically in a browser how to deal effectively with conflicting dependencies.  npm and CommonJS modules make big assumptions about having unfettered access to a filesystem, which browsers do not.  npm3 "helps" in its default behaviour of flattening dependencies.  Many people don't fully understand or appreciate the semantics of npm (including myself sometimes, as it was only relatively recently I began to understand `peerDependencies` and I had to help @dylans understand why that was better for a lot of things).

For our own purposes, a "bare metal" [npm toolchain](http://blog.keithcirkel.co.uk/why-we-should-stop-using-grunt/) might be fine, but it takes a fair amount of knowledge to be able to build something incremental on top of it.  My feeling is that we should have a system that allows people to easily integrate our packages into their preferred toolchain (which is why having them on npm and hopefully bower makes that the easiest way for them to be consumed) but also provide an "out of the box" toolchain that works.  This provides an example of how we intend the packages to be consumed as well as provides those without a toolchain something to work with.  In my opinion, one of the biggest objectives we need around this space is that if a new developer cannot get a full Dojo 2 app running in less than 15 minutes, we have failed.

> I appreciate that we don't want to ask people to install 20 packages in order to build a Dojo 2 application, but then why even have dojo/compose outside of dojo/core?

We will always be striking a balance between `left-pad`/`lodash` and `dojo/dojo`.  At the most basic, a Dojo 2 package meets these criteria:
- A clear concise functional purpose
  - That logically could have its own development/release cycle
  - And could be managed with a maintainer
- No circular dependencies on other packages
- Minimise runtime dependencies on other packages

**Comment by kitsonk** [comment]
Attempting to pull all the shim code into [dojo/shim](https://github.com/dojo/shim).

I have also been able to address handing off to native a bit better than we we are currently doing in `dojo/core`.

**Comment by kfranqueiro** [comment]
It's unclear to me how this evolved from:

> the functional enhancements in core and the standards based shims are not easily separable

to:

> Attempting to pull all the shim code into dojo/shim.

Can you elaborate on how the path was "cleared" for this?

Also, I'm at least a little concerned about all of the original core shims' authorship being completely nuked from orbit in the new repo, but I'm not sure it'd be terribly straightforward to preserve.

**Comment by kitsonk** [comment]
> Can you elaborate on how the path was "cleared" for this?

For now, quantifying that up and seeing what we are left with is what is underway.  They appear to be more easily separable than I had originally assumed.  It is worth taking a look at, as we continue to find ourselves in the situation of "dragging" around polyfills that are not needed in most browsers, but it is increasingly looking like loader plugins won't be part of ES Module Loading, which means that run-time conditional loading of modules will be challenging.  It also doesn't necessarily fit into a build pipeline.  Having the shims broken out give us an opportunity to create modules that can be substituted that simply re-export the native functions.  Including those "light" re-exported modules would be the plan to be part of the module.  Once we quantify up exactly what is required to be "duplicated" between `shim` and `core` we can figure out how best to provide that overlap.

> Also, I'm at least a little concerned about all of the original core shims' authorship being completely nuked from orbit in the new repo, but I'm not sure it'd be terribly straightforward to preserve.

Nor was `dojo/dojo` to `dojo/dojo2` to `dojo/core`...  What is the concern?  All the code was contributed under the appropriate licenses and under a CLA.

**Comment by novemberborn** [comment]
> Also, I'm at least a little concerned about all of the original core shims' authorship being completely nuked from orbit in the new repo, but I'm not sure it'd be terribly straightforward to preserve.

Could fork `dojo-core` and rewrite the history (`git filter-branch`) to remove the non-shim files. Still won't be easy though.

**Comment by kitsonk** [comment]
> Also, I'm at least a little concerned about all of the original core shims' authorship being completely nuked from orbit in the new repo, but I'm not sure it'd be terribly straightforward to preserve.

I forked core and ported over all the changes everything I did today in regards to `dojo/shim` did a forced push.  This preserves the history/blame/etc. of anything that is remaining in the repo.
